src/sv-pipeline/04_variant_resolution/scripts/clean_vcf_part5_update_records.py

```diff
@@ -0,0 +1,185 @@
+
+    if args.sexchr_revise.endswith(".gz"):
+        sexchr_revise = set([line.rstrip() for line in gzip.open(args.sexchr_revise, 'rt')])
+    else:
+        sexchr_revise = set([line.rstrip() for line in open(args.sexchr_revise, 'rt')])
+    print("{} sites to revise on sex chromosomes".format(len(sexchr_revise)), file=sys.stderr)
+
+    if args.multi_geno_ids_txt.endswith(".gz"):
+        multi_geno_ids = set([line.rstrip() for line in gzip.open(args.multi_geno_ids_txt, 'rt')])
+    else:
+        multi_geno_ids = set([line.rstrip() for line in open(args.multi_geno_ids_txt, 'rt')])
+    print("{} multiallelic sites".format(len(multi_geno_ids)), file=sys.stderr)
+
+    normal_vcf = pysam.VariantFile(args.normal_revise_vcf)
+
+    NEW_HEADER_LINES = ['##ALT=<ID=CNV,Description="Copy Number Polymorphism">',
+                        '##FORMAT=<ID=CNQ,Number=1,Type=Integer,Description="Read-depth genotype quality">',
+                        '##FORMAT=<ID=CN,Number=1,Type=Integer,Description="Predicted copy state">',
+                        '##FILTER=<ID=PESR_GT_OVERDISPERSION,Description="High PESR dispersion count">',
+                        '##FILTER=<ID=MULTIALLELIC,Description="Multiallelic site">']
+
+    # # Add metadata lines for annotations
+    header1 = normal_vcf.header
+
+    for f in NEW_HEADER_LINES:
+        header1.add_line(f)
+        header2.add_line(f)
+
+    non_outlier_samples = set([s for s in header1.samples if s not in outlier_samples])
+    vf_1 = max(len(non_outlier_samples) * 0.01, 2)
+
+    biallelic_gts = {(1, 1), (0, 0), (0, 1), (None, None)}
+
+    print("reformatting records", file=sys.stderr)
+    cleangq_filename = "cleanGQ.vcf.gz"
+    cleanqg_out = pysam.VariantFile(cleangq_filename, 'w', header=normal_vcf.header)
+
+    mulitallelic_filename = "multiallelic.vcf.gz"
+    multiallelic_out = pysam.VariantFile(mulitallelic_filename, 'w', header=normal_vcf.header)
+
+    no_variant_samples_list_file = "no_called_samples.list"
+    no_variant_samples_out = open(no_variant_samples_list_file, 'w')
+
+    for idx, record in enumerate(normal_vcf):
+        multi_del = False
+        multi_dup = False
+        gt4_copystate = False
+        gt5kb_dup = False
+        gt5kb_del = False
+        if (idx - 1) % 1000 == 0:
+            print("processed {} records".format(idx), file=sys.stderr)
+        if record.id in revised_lines_by_id:
+            record = revised_lines_by_id[record.id]
+        if record.info.get('SVTYPE', None) == 'DEL':
+            if abs(record.stop - record.pos) >= 1000:
+                sample_cn_map = {s: record.samples[s]['RD_CN'] for s in non_outlier_samples}
+                if len([s for s in sample_cn_map if (sample_cn_map[s] is not None and sample_cn_map[s] > 3)]) > vf_1:
+                    multi_del = True
+            gts = [record.samples[s]['GT'] for s in non_outlier_samples]
+            if len([gt for gt in gts if gt not in biallelic_gts]) > 0:
```

Comment:
```suggestion
            if any(gt not in biallelic_gts for gt in gts):
```
Works with generator so faster, avoids memory alloc and dealloc.